### PR TITLE
Tag block

### PR DESCRIPTION
**Pull request by sagikazarmark**
Implemented a tags block. It can be configured to use different contexts (although currently it can only accept a string). Number of displayed tags can also be configured, but limitation is disabled by default.

### CODE REVIEW

**Comment by sagikazarmark** [comment]
ping @rande

File: Block/TagsBlockService.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
+ */
+class TagsBlockService extends BaseBlockService
+{
+    protected $manager;
```
Comment (OskarStark) [inline Block/TagsBlockService.php]: can you please add

```php
/**
 * @var TagManagerInterface
 */
```

File: Block/TagsBlockService.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                    'acceptNull' => false,
+                ))
+            ->end()
+            ->with('settings.context')
+                ->assertNotBlank()
+            ->end()
+            ->with('settings.title')
+                ->assertNotBlank()
+                ->assertMaxLength(array('limit' => 50))
+            ->end();
```
Comment (OskarStark) [inline Block/TagsBlockService.php]: can you put the ``;`` to the next line, like the generated admin classes do with their mappers?

File: Block/TagsBlockService.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildEditForm(FormMapper $formMapper, BlockInterface $block)
+    {
+        $formMapper->add('settings', 'sonata_type_immutable_array', array(
```
Comment (OskarStark) [inline Block/TagsBlockService.php]: would prefer:

```php
    $formMapper
        ->add() // ...
    ;
```

File: Resources/config/block.xml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="sonata.classification.block.tags" class="Sonata\ClassificationBundle\Block\TagsBlockService">
+            <tag name="sonata.block"/>
```
Comment (OskarStark) [inline Resources/config/block.xml]: missing space before ```/>```

File: Resources/translations/SonataClassificationBundle.de.xliff [hunk 5]
```diff
@@ -162,6 +162,10 @@
                 <source>breadcrumb.link_category_tree</source>
                 <target>Kategorie</target>
             </trans-unit>
+            <trans-unit id="no_tag_found">
+                <source>no_tag_found</source>
+                <target>No tag found.</target>
```
Comment (OskarStark) [inline Resources/translations/SonataClassificationBundle.de.xliff]: german translation is: ```Kein Tag gefunden.```

File: Resources/translations/SonataClassificationBundle.es.xliff [hunk 6]
```diff
@@ -158,6 +158,10 @@
                 <source>classification.tree_mode</source>
                 <target>classification.tree_mode</target>
             </trans-unit>
+            <trans-unit id="no_tag_found">
+                <source>no_tag_found</source>
+                <target>No tag found.</target>
```
Comment (OskarStark) [inline Resources/translations/SonataClassificationBundle.es.xliff]: If we don't have a translation, please use the source key for the target in this case:

```xml
<trans-unit id="no_tag_found">
    <source>no_tag_found</source>
    <target>no_tag_found</target>
</trans>
```

File: Resources/views/Block/tags.html.twig [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+                <i class="fa fa-tags"></i> {{ settings.title }}
+            </h3>
+        </div>
+
+        <div class="box-body">
+            {% sonata_template_box 'This is the tags block.' %}
+
+            <ul class="list-group">
+                {% for tag in pager.getResults() %}
+                    <li><a class="list-group-item" href="{{ url('sonata_news_tag', { 'tag': tag.name }) }}">{{ tag.name }}</a></li>
```
Comment (OskarStark) [inline Resources/views/Block/tags.html.twig]: would prefer a linebreak after the ```<li>``` node

File: Resources/views/Block/tags.html.twig [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+                {% else %}
+                    <li class="list-group-item" href="#">{{ 'no_tag_found'|trans({}, 'SonataClassificationBundle') }}</li>
```
Comment (OskarStark) [inline Resources/views/Block/tags.html.twig]: ```href``` on ```<li>``` ?

**Comment by OskarStark** [comment]
after merge an implementation in the sandbox would be cool :+1:

File: Resources/translations/SonataClassificationBundle.fr.xliff [hunk 9]
```diff
@@ -154,6 +154,10 @@
                 <source>classification.tree_mode</source>
                 <target>Arbre</target>
             </trans-unit>
+            <trans-unit id="no_tag_found">
+                <source>no_tag_found</source>
+                <target>No tag found.</target>
```
Comment (Soullivaneuh) [inline Resources/translations/SonataClassificationBundle.fr.xliff]: "Aucun tag trouvé" for french translation. :+1:

File: Block/TagsBlockService.php (hunk 1 above)
Comment (sagikazarmark) [inline Block/TagsBlockService.php]: Added, but it's omitted everywhere else. Is there a standard for that?

Comment (OskarStark) [inline Block/TagsBlockService.php]: i'm not sure

File: Block/TagsBlockService.php (hunk 3 above)
Comment (sagikazarmark) [inline Block/TagsBlockService.php]: Added, also copied code.

File: Block/TagsBlockService.php (hunk 1 above)
Comment (Soullivaneuh) [inline Block/TagsBlockService.php]: The standard is to add it.

> but it's omitted everywhere else

Well, was an historic mistake, time to change. :-)

File: Resources/views/Block/tags.html.twig (hunk 8 above)
Comment (OskarStark) [inline Resources/views/Block/tags.html.twig]: in the first the class ```list-group-item``` is on the ```<a>``` and in the second, the css class is on the ```<li>```, is that correct?

**Comment by sagikazarmark** [comment]
I added everywhere the English form of the translation. Is that right? I will update German and French ones.

**Comment by sagikazarmark** [comment]
Just saw @OskarStark's comment about translation files. Pushed fixes. Please review once again, then I squash.

**Comment by Soullivaneuh** [comment]
@sagikazarmark You ignore my comment, I'm so sad. :'(

https://github.com/sonata-project/SonataClassificationBundle/pull/143#discussion_r40887597

**Comment by sagikazarmark** [comment]
@Soullivaneuh sorry, ignore what?

**Comment by Soullivaneuh** [comment]
> "Aucun tag trouvé" for french translation. :+1:

**Comment by sagikazarmark** [comment]
I still don't understand. Why should I ignore it? Is it a bad translation?

**Comment by Soullivaneuh** [comment]
Sorry I would say "You ignor**ed**" my comment, was a joke. :-P

The translation is good, but you didn't put it on your PR: https://github.com/sonata-project/SonataClassificationBundle/pull/143/files#diff-aa46db5c4f62661c67b3f0a1e5d0ab58R159

**Comment by sagikazarmark** [comment]
Oh, sorry, I thought I updated the french translation too. Now it can get it's own commit. Are you happy? :wink:

**Comment by Soullivaneuh** [comment]
\o/

File: Block/TagsBlockService.php (hunk 1 above)
Comment (OskarStark) [inline Block/TagsBlockService.php]: we should change this everywhere IMO

@sagikazarmark could you please create a ticket in the ```SonataAdminBundle```  for that, and describe what do to?

maybe we should list all sonata bundles and mark them as checked if we did the work

**Comment by OskarStark** [comment]
and could you please provide a screenshot of the block?
Maybe when it has tags, and when its empty?

**Comment by sagikazarmark** [comment]
It looks exactly the same as the Recent Posts block. But I can set it up in sandbox and send a PR there as well. (But to make that work, we have to merge this first)

File: Block/TagsBlockService.php (hunk 1 above)
Comment (Soullivaneuh) [inline Block/TagsBlockService.php]: > we should change this everywhere IMO

Maybe, but not in this PR.

**Comment by OskarStark** [comment]
IMO this PR is ready to merge.

@Soullivaneuh push the button if you like :smile:

**Comment by sagikazarmark** [comment]
Squash first?

**Comment by OskarStark** [comment]
yes please

**Comment by sagikazarmark** [comment]
Done

**Comment by Soullivaneuh** [comment]
I'm not using this bundle for my projects.

And especialy because of this, screenshots and documentation update should be put on this PR.

@sagikazarmark could you please handle it?

**Comment by sagikazarmark** [comment]
Sure thing.

**Comment by sagikazarmark** [comment]
Actually I cannot see any screenshots nor documentation for most of blocks. How should I proceed? Where to put documentation? Classification -> Blocks? Where to put images?

Shouldn't it be in a separate PR where docs is added for all other blocks as well? Since there is no base documentation.

**Comment by Soullivaneuh** [comment]
> Actually I cannot see any screenshots nor documentation for most of blocks. How should I proceed? 

Docs: https://github.com/sonata-project/SonataClassificationBundle/tree/master/Resources/doc/reference

For screenshots, just add them to your PR description.

**Comment by OskarStark** [comment]
would add ```block_tags.rst``` to ```Resources/doc/reference/```

**Comment by OskarStark** [comment]
ping @sagikazarmark

**Comment by sagikazarmark** [comment]
Sorry, will deal with this in the weekend.

**Comment by OskarStark** [comment]
:+1: thank you!

**Comment by OskarStark** [comment]
ping @sagikazarmark can you please fix the docs, rebase the branch and add a screenshot of the block?

this would be awesome :+1:

File: Resources/views/Block/tags.html.twig [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            </h3>
+        </div>
+
+        <div class="box-body">
+            {% sonata_template_box 'This is the tags block.' %}
+
+            <ul class="list-group">
+                {% for tag in pager.getResults() %}
+                    <li class="list-group-item">
+                        <a href="{{ url('sonata_news_tag', { 'tag': tag.name }) }}">{{ tag.name }}</a>
```
Comment (core23) [inline Resources/views/Block/tags.html.twig]: Is there a reason why you generate a route to the news?

Comment (sagikazarmark) [inline Resources/views/Block/tags.html.twig]: I don't understand the question. What should I do instead?

Comment (core23) [inline Resources/views/Block/tags.html.twig]: You generate a route to ``sonata_news_tag``, but this should be a general route which has no dependency to the ``NewsBundle``.

Comment (sagikazarmark) [inline Resources/views/Block/tags.html.twig]: Oh, you are right. Good catch. Any suggestions how this view should be changed? AFAIK there are no tags classification controller.

Comment (core23) [inline Resources/views/Block/tags.html.twig]: I've created something similar for my personal use. 

An abstract ``TagController`` in the ``ClassificationBundle`` and the implementation in the different bundles which only overrides the ``template`` and the ``context``

Comment (sagikazarmark) [inline Resources/views/Block/tags.html.twig]: Sounds like a good idea. @OskarStark objections?

Comment (OskarStark) [inline Resources/views/Block/tags.html.twig]: sounds good to me

Comment (core23) [inline Resources/views/Block/tags.html.twig]: @sagikazarmark Do you want to work on it or should I propose my solution?

Comment (sagikazarmark) [inline Resources/views/Block/tags.html.twig]: Currently there are a few things on my plate, so I can't really play with it. I can add you a collaborator to my fork, so you don't have to clone/fork/repush, etc.

Comment (sagikazarmark) [inline Resources/views/Block/tags.html.twig]: If no work is done on this then I am picking it up. After thinking about this, I think the abstract TagController is not so good idea after all.

What if we just add a route option to the block?

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Comment by sagikazarmark** [comment]
@core23 Have you took care of this or should I rework and resubmit it?

**Comment by core23** [comment]
I'll rebase my PR @sagikazarmark

**Pull request closed without merging**